### ISSUE TITLE
Remove dependency matching

### ISSUE THREAD

**Issue by robertvazan** [appc/spec#295]
Apparently the use case for dependency matching is to allow security patches for the operating system to be deployed without having all app developers publish updated versions of their packages. This however won't work, because typical image build works like this:

1. Take base image,
2. install frameworks and libraries, and
3. add application files.

While dependency matching might allow (1) to receive security patches, it won't help with (2). And step (2) is when sensitive stuff like web server gets installed. Even the patches for (1) are likely to break something, because apps will customize the system they run on.

I think the dependency matching has no use case and it just makes for fragile images that are bound to fail. I suggest removing it and going for compulsory identification of dependencies via hashes. Nobody is going to edit image manifests by hand anyway. That's the job of build system.

In other words, I propose strict separation between build and images. Images should be immutable. They must not change with updates. That's why they have checksum after all. Registries can still implement dependency matching, because build systems can make use of that. I just think that deployment system is the worst place to make decisions as to which base image to use.

As for letting security patches to be applied smoothly, I think there's no other choice than to rebuild all images. Perhaps AppContainer could encourage this practice by providing means to run standardized sandboxed builds on the target system. Something like Dockerfile or maybe something like executable build images that produce or turn themselves into immutable images with latest patches applied.

**Comment by philips** [comment]
On Fri, Apr 10, 2015 at 5:25 PM, Robert Važan <notifications@github.com> wrote:
>
> Apparently the use case for dependency matching is to allow security patches for the operating system to be deployed without having all app developers publish updated versions of their packages. This however won't work, because typical image build works like this:
>
> Take base image,
> install frameworks and libraries, and
> add application files.
>
> While dependency matching might allow (1) to receive security patches, it won't help with (2). And step (2) is when sensitive stuff like web server gets installed. Even the patches for (1) are likely to break something, because apps will customize the system they run on.
>
> I think the dependency matching has no use case and it just makes for fragile images that are bound to fail. I suggest removing it and going for compulsory identification of dependencies via hashes. Nobody is going to edit image manifests by hand anyway. That's the job of build system.

Image builders have the option to use explicit image IDs but they must
provide a name too so we can figure out where to download the image
from. I see no reason to limit the spec by forcing builders to provide
an ID. If the builder does not want to risk breakage and wants a 100%
dependable system they should use image IDs.

Does this seem reasonable?

**Comment by robertvazan** [comment]
I agree that name is necessary to locate the server that provides the image. I noted it in #296 too. Nevertheless any kind of dependency matching via labels/versions on target host is undesirable and unnecessary.

"You don't have to use it" is not a good argument to support a feature. Devs will be lazy and just reference "latest" tag. Why not stop the impeding disaster now and ban the practice? When problems occur, people will say "ACIs are fragile", not "devs are stupid".

Common standard like AppContainer should be lean and mean. Otherwise we risk opening pandora's box with other build-related features. "Cut-cut-cut" should be the matra of anyone working on standards. If it isn't required for interoperability, then just nix it.

**Comment by jonboulle** [comment]
> Nevertheless any kind of dependency matching via labels/versions on target host is undesirable and unnecessary.

why? Are you coming at this from the perspective that users never interact with ACIs?
I agree with you from the purist perspective - in the ideal scenario, humans aren't involved and this is completely integrated with build and orchestrated systems - but I'm not sure why it needs to be quite so restrictive in the spec at this stage. It seems like more of a policy issue.

**Comment by robertvazan** [comment]
I am not trying to implement policy at tool level. My concern is merely with complexity of the spec. We already have quite a few "run anywhere" platforms, most notably java, but the complexity of these platforms makes apps hard to deploy in practice. The point of containers is that the interface is bare minimum, thus compatibility issues are very unlikely and wider selection of tools is available. Dependency matching is redundant on image level (there's no use case that could be recommended practice), thus it's prime candidate for removal from the spec.

**Comment by robertvazan** [comment]
Perhaps the requirement to have moving underlying image might be more generally addressed by some kind of "source" images (or self-extracting/self-downloading images) that run through well-defined build process to generate target images. But then I still think this is best done via external build/orchestration tools.

**Comment by robertvazan** [comment]
Also note that mere presence of this feature might cripple some advanced tools that automatically optimize image size by merging multiple parent images. This is useful when several components of the image (typically OS, framework, and the app) update independently of each other. Such multi-parent images can weight well under 1MB, permitting archival of every single version of the image ever released. Image optimization can be reliably performed only if the image is fully defined, i.e. there's no deploy-time image lookup/matching. I was even going to suggest Mercle tree hashing to support transparent image graph optimization.

**Comment by globalcitizen** [comment]
As the author of #394 and #381 I strongly support @robertvazan 's perspective here, both in terms of a minimalist mindset and in terms of placing determinism as a first class citizen in reconsidering the current spec.

However, I would even go a step further and suggest removing name-based image downloads as a needless source of bugs through avoidable non-determinism. Yes, that means, it should not be possible to have the network as an input to a random DNS-based (or anything based) deployment process.

Reasoning: Why go to the hassle of specifying a half-baked download standard when you can just require a file with a certain hash and leave the secure transfer out of scope? This also removes an internet network dependency for ACE implementations... always a good thing when running code from somewhere else from a security standpoint.

**Comment by vbatts** [comment]
I will not argue that there isn't a need for examples and patterns, but there is an absolute use case.
When getting into ISV deployment, there will arise the case of non - re distribution of their software, and that the ISV's software was built for an explicit version of base image, that they too are unable to redistribute.

These stacked dependencies have the ability to be signed and trusted (even encrypted), fully qualified names and content addressed. 

Flattening images and optimizing is all great, but blocking this pattern will put users in the position of violating commercial software terms and agreements, and loosing the isolated trust of each stack dependency.

**Comment by globalcitizen** [comment]
I don't think you understood my point - when I suggested removing "name-based image downloads" owing to its destruction of determinism, I wasn't merely criticizing the name-based notion ("fully qualified" or otherwise) but in fact the image download notion in general.

As far as justifying that feature on an obscure legal edge-case, well, I don't feel that has much weight in pain saved versus everyone else using open source or in-house stuff which we all know will be 99.99% of use.
